---
title: "EDA"
author: "Ted Ladas - s2124289"
date: "11/07/2021"
output: html_document
---

# Libraries and Requirements

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse) # for tidy syntax
library(janitor) # for handling column names
library(data.table) # for fast processing of data
library (skimr) # to get a fast grasp of the dataset
library(broom) 
library(tidytext)
library(patchwork) 
library(corrplot)
library(caret)
library(mice)
```

# Load the data files

```{r loading}
keys <-
  as.data.table(read_csv("./data/crypto_wake_up_sample_sheet.csv")) %>% clean_names()
genes <-
  as.data.table(
    read_csv(
      "./data/H99_all_genes_promoter_500nt_4mer_counts.csv",
      skip = 10,
      col_names = FALSE
    ) %>%
      row_to_names(row_number = 1)
  ) %>% clean_names() %>% mutate_at(vars(-gene), as.numeric)
experiment <-
  as.data.table(read_csv(
    "./data/CW_kallisto_abundance_fold_change_long_by_gene.csv"
  )) %>% clean_names()
```

# Missing Data Handling

```{r help-missing}
handle.missing.data <- function(df) {
  df <- df %>%
    mutate(fold_change = case_when(est_counts + tpm == 0 ~ 0,
                                   TRUE ~ fold_change)) %>% 
    mutate(fold_change = na_if(fold_change, Inf))
  return(df)
}
```

```{r missing}
experiment <- experiment %>% handle.missing.data()
```

```{r load-check}
keys %>% head() 
genes %>% head() 
experiment %>% head() 
```

# Joining Datasets

```{r joins}
exp <- experiment %>% 
  left_join(keys, by = "code") %>%
  select(gene, medium, temp, time, rep, est_counts, tpm, fold_change)
skim(exp)

df <- exp %>%
  left_join(genes, by = "gene") %>%
  mutate(row_name = paste(gene, medium, temp, time, rep, sep="-"), .before=gene) %>%
  remove_rownames %>%
  column_to_rownames("row_name") %>% as.data.frame()
df %>% head() 
```

```{r model-data}
dmy <- dummyVars("~ medium + temp + time + rep ", data = df)
df.model <- data.frame(predict(dmy, newdata = df))
df.model <- cbind(df.model, df) %>%
  select(-medium, -temp, -time, -rep, -gene) %>% 
  clean_names() %>% 
  relocate(c(est_counts, tpm, fold_change), .before=medium_rpmi)

df.model %>% head() 
```

# Univariate Analysis

```{r help-uni}
plot.univariate <- function(var, density, dat = exp) {
  color = "blue4"
  density = TRUE
  if (var == "est_counts") {
    density = FALSE
  }
  
  if (density) {
    gg.var <- ggplot(data = dat, aes(x = .data[[var]])) +
      geom_density(col = color,
                   fill = color,
                   alpha = 0.8) +
      theme(axis.title.x = element_blank(),
            axis.title.y = element_blank()) +
      ggtitle(var) +
      xlim(0, 10 + 90 * (var != "fold_change")) +
      theme_minimal()
  } else {
    gg.var <- ggplot(data = dat, aes(x = .data[[var]])) +
      geom_histogram(
        col = color,
        fill = color,
        alpha = 0.7,
        bins = 30
      ) +
      theme(axis.title.x = element_blank(),
            axis.title.y = element_blank()) +
      ggtitle(var) +
      xlim(0, 10 + 90 * (var != "fold_change")) +
      theme_minimal()
  }
  return(gg.var)
}
```

```{r plot-uni}
names <- c("est_counts", "tpm", "fold_change")
names <- set_names(names)
plots <- map(names, ~ plot.univariate(.x))
plots[[1]] + plots[[2]] + plots[[3]]
```

Let's take fold_change as the y variable.

```{r densities}
df %>% select(gene, medium, temp, time, rep, fold_change) %>%
  ggplot(aes(x = fold_change, col = medium, fill = medium), show.legend = FALSE) +
  geom_density(alpha = 0.8) +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank()) +
  ggtitle("TPM") +
  facet_grid(medium ~ .) +
  xlim(0, 10) +
  theme_minimal() +
  theme(legend.position = "none")
```

# PCA on full data

## Centerint tha data matrix

```{r centering}
df.exp <- df.model
df.sample <- df.exp %>% sample_n(size=nrow(df.exp)/2, replace=FALSE) %>% filter(fold_change<5)

df.X <- df.sample %>% select(-est_counts, -tpm, -fold_change)
df.y <- df.sample %>% select(fold_change)

df.cntr.X <- select_if(df.X, is.numeric) %>% 
  scale(center = TRUE, scale = TRUE) %>% 
  as.data.table()
df.cntr.X %>% head() 

svd.results <- svd(df.cntr.X)
```

## Variance explained by PC score

```{r var-expl}
# interpret variance explained 
svd.var.exp <- tidy(svd.results, matrix = "d")

svd.var.exp %>% ggplot(aes(x = PC, cumulative)) +
  geom_line() +
  # scale_x_continuous(breaks = seq(min(svd.var.exp$PC),max(svd.var.exp$PC),by=1)) +
  # coord_cartesian(xlim = c(1, 200)) +
  labs(title = "Variance Explained by PCs") +
  theme_minimal()
```

## Column wise interpretation of PCA loadings

```{r col-expl}
# interpret loadings - working on columns
svd.loadings <- tidy(svd.results, matrix="v") %>%
                mutate(element = colnames(df.cntr.X)[column])

svd.loadings %>% filter(PC<=4) %>%
  group_by(PC) %>%
  top_n(10, abs(value)) %>%
  ungroup() %>%
  mutate(element = reorder_within(element, value, PC)) %>%
  ggplot(aes(x=element, y=value, fill=factor(PC))) +
  geom_col(show.legend = FALSE) +
  facet_wrap(~PC, scales="free") +
  scale_x_reordered() + 
  coord_flip() +
  labs(title = "First four PCs of feautures") + 
  theme_minimal()
```

## Row wise interpretation of PCA loadings

```{r row-expl}
# interpret loadings - working on rows
svd.loadings.row <- tidy(svd.results, matrix="u") %>%
                mutate(gene_names = rownames(df.X)[row])

svd.loadings.row %>% filter(PC<=4) %>%
  group_by(PC) %>%
  top_n(10, abs(value)) %>%
  ungroup() %>%
  mutate(gene_names = reorder_within(gene_names, value, PC)) %>%
  ggplot(aes(x=gene_names, y=value, fill=factor(PC))) +
  geom_col(show.legend = FALSE) +
  facet_wrap(~PC, scales="free") +
  scale_x_reordered() + 
  coord_flip() +
  labs(title = "First four PCs of feautures") + 
  theme_minimal()
```

## PCA Biplot

```{r biplot}
# final plot
svd.pcs <- tidy(svd.results, matrix = "u")

svd.pcs.1 <- svd.pcs %>% filter(PC == 1) %>%
  mutate(PC1 = value) %>%
  select(PC1)
svd.pcs.2 <- svd.pcs %>% filter(PC == 2) %>%
  mutate(PC2 = value) %>%
  select(PC2)
svd.pcs.3 <- svd.pcs %>% filter(PC == 3) %>%
  mutate(PC3 = value) %>%
  select(PC3)
df.pcs <- cbind(df.y, svd.pcs.1, svd.pcs.2, svd.pcs.3)

df.pcs %>%
  ggplot(aes(
    x = PC1,
    y = PC2,
    col = fold_change,
    fill = fold_change
  )) +
  geom_jitter(shape = 21, col = "black") +
  labs(title = "PC1 vs PC2", fill = "fold_change") +
  theme_minimal()
```

# PCA on experiment data

## Centerint tha data matrix

```{r centering-2}
df.exp <- df.model 
df.sample <- df.exp %>% sample_n(size=nrow(df.exp)/2, replace=FALSE) %>% filter(fold_change<5) %>% select(1:10)

df.X <- df.sample %>% select(-est_counts, -tpm, -fold_change)
df.y <- df.sample %>% select(fold_change)

df.cntr.X <- select_if(df.X, is.numeric) %>% 
  scale(center = TRUE, scale = TRUE) %>% 
  as.data.table()
df.cntr.X %>% head() 

svd.results <- svd(df.cntr.X)
```

## Variance explained by PC score

```{r var-expl-2}
# interpret variance explained 
svd.var.exp <- tidy(svd.results, matrix = "d")

svd.var.exp %>% ggplot(aes(x = PC, cumulative)) +
  geom_line() +
  # scale_x_continuous(breaks = seq(min(svd.var.exp$PC),max(svd.var.exp$PC),by=1)) +
  # coord_cartesian(xlim = c(1, 200)) +
  labs(title = "Variance Explained by PCs") +
  theme_minimal()
```

## Column wise interpretation of PCA loadings

```{r col-expl-2}
# interpret loadings - working on columns
svd.loadings <- tidy(svd.results, matrix="v") %>%
                mutate(element = colnames(df.cntr.X)[column])

svd.loadings %>% filter(PC<=4) %>%
  group_by(PC) %>%
  top_n(10, abs(value)) %>%
  ungroup() %>%
  mutate(element = reorder_within(element, value, PC)) %>%
  ggplot(aes(x=element, y=value, fill=factor(PC))) +
  geom_col(show.legend = FALSE) +
  facet_wrap(~PC, scales="free") +
  scale_x_reordered() + 
  coord_flip() +
  labs(title = "First four PCs of feautures") + 
  theme_minimal()
```

# Outlier examples

```{r outlier-identifier}
df.full <- df %>%
  select(gene, medium, temp, time, rep, fold_change) %>%
  filter(rep=="A") %>% 
  select(-rep) %>% 
  group_by(gene, medium, temp) %>%
  filter(time %in% c(10, 120)) %>%
  ungroup()

df10 <- df.full %>% filter(time==10)
df120 <- df.full %>% filter(time==120)

df.format <- df10 %>%
  left_join(df120, by=c("gene", "medium", "temp")) %>%
  mutate(fold_change.10=fold_change.x, .before=fold_change.x) %>%
  mutate(fold_change.120=fold_change.y, .before=fold_change.y) %>%
  select(-time.y, -time.x, -fold_change.x, -fold_change.y) 

extreme.cases <- df.format %>% 
  group_by(gene, medium, temp) %>% 
  mutate(fold_change_diff = fold_change.120 - fold_change.10) %>% 
  arrange(-fold_change_diff) %>% 
  ungroup() %>%
  filter(row_number() > max(row_number()) - 5 | row_number() <= 5)

cases <- extreme.cases %>% select(gene) %>% as.matrix()
```

```{r time-series}
piazza <- c("CNAG_00121", "CNAG_06917")
pca.example <- c("CNAG_01842", "CNAG_03788")

big <- exp %>%
  filter(gene %in% cases[1:5], rep == "A", temp != "30C") %>%
  select(-est_counts, -tpm, -rep) %>%
  ggplot(aes(x = time, y = fold_change, col=gene)) +
  geom_line() +
  geom_point() +
  theme_minimal() +
  theme(legend.position = "none") +
  labs(title="Most overall change over time",
       subtitle=cases[1:5]) +
  facet_wrap( ~ medium + temp, scale="free") 

small <- exp %>%
  filter(gene %in% cases[6:10], rep == "A", temp != "30C") %>%
  select(-est_counts, -tpm, -rep) %>%
  ggplot(aes(x = time, y = fold_change, col=gene)) +
  geom_line() +
  geom_point() + 
  theme_minimal() +
  theme(legend.position = "none",
        axis.title.y=element_blank()) +
  labs(title="Least overall change over time",
       caption=paste("Most: ", paste(cases[1:5], collapse = ', '), "\n",
                     "Least: ", paste(cases[6:10], collapse = ', '))) +
  facet_wrap( ~ medium + temp, scale="free") 

big + small
```

# Bivariate Analysis

```{r bivar}
df <- df %>% as.data.table()
numcols <- sapply(df, is.numeric)
df.corr <- df[, ..numcols] %>% cor(use="pairwise.complete")

df.corr.big <- df.corr*(abs(df.corr)>0.6)
df.corr.big <- df.corr.big[-which(abs(rowSums(df.corr.big))==1),-which(abs(colSums(df.corr.big))==1)]

square.cor <- corrplot(df.corr.big,
                       order="hclust",
                       addrect=14,
                       method='square',
                       diag=FALSE,
                       tl.col="black",
                       tl.cex = 0.9,
                       outline=TRUE,
                       cl.lim=c(-1, 1),
                       mar=c(0,0,1.5,0))
```

# Question 1

Find genes that have different expression across design conditions.

```{r question-1}
df
```

# Question 2

Find motifs that are associated with genes that have different expression across design conditions (and are not associated with the ones that are not differentially expressed).

Investigate whether a (relatively) small set of motifs can predict the observed patterns of gene expression.

```{r question-2}
experiment %>% select(est_counts, tpm, fold_change) %>% md.pattern()

experiment %>%
  select(est_counts, tpm, fold_change) %>%
  mice(printFlag=FALSE, seed=1903, m=5, method="pmm") %>%
  stripplot()

imps <- experiment %>%
  select(est_counts, tpm, fold_change) %>%
  mice(printFlag=FALSE, seed=1903, m=10, method="pmm")
fit <-  with(data=imps, exp=lm(fold_change ~ est_counts + tpm))
pooled <- pool(fit)

pooled_lm = fit$analyses[[1]]

```
